File: 3.services/src/service_py_pkg/service_py_pkg/acc_client.py
#!/usr/bin/env python3
from functools import partial

import rclpy
from rclpy.node import Node

from custom_interfaces.srv import Acc


class AccClient(Node):
    """A simple ROS2 node that creates a client for the service "accumulate" and accumulates the received numbers."""

    def __init__(self):
        super().__init__("acc_client_node")
        self.get_logger().info("Accumulate Client Node has been started!")
        # Create a client
        self.cli = self.create_client(Acc, "accumulate")

    def add(self, a: int, b: int, c: int):
        """Method to call the service "accumulate" with the given numbers."""
        # Wait for the service to be available
        while not self.cli.wait_for_service(timeout_sec=1.0):
            self.get_logger().info("Service not available, waiting again...")
        # Create a request
        req = Acc.Request()
        req.a = a
        req.b = b
        req.c = c
        # Call the service asynchronously
        future = self.cli.call_async(req)
        # Add a callback to handle the response
        # Use partial to pass the arguments to the callback
        future.add_done_callback(partial(self.response_callback, request=req))

    def response_callback(self, future, request):
        """Callback function for the service response."""
        try:
            response = future.result()
            self.get_logger().info(
                f"Result of {request.a} + {request.b} + {request.c}: {response.sum}"
            )
        except Exception as e:
            self.get_logger().error(f"Service call failed: {e}")


def main(args=None):
    # Initialize the ROS2 Python client library
    rclpy.init(args=args)

    # Create a ROS2 node, then spin the node to keep it alive, finally destroy the node
    node = AccClient()
    node.add(1, 2, 3)
    node.add(4, 5, 6)
    node.add(7, 8, 9)
    rclpy.spin(node)
    node.destroy_node()

    # Shutdown the ROS2 Python client library
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: 3.services/src/service_cpp_pkg/src/acc_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "custom_interfaces/srv/acc.hpp"

class AccClientNode : public rclcpp::Node
{
public:
    AccClientNode() : Node("acc_client_node")
    {
        // Create a client for the "accumulate" service
        client_ = this->create_client<custom_interfaces::srv::Acc>("accumulate");
    }
    rclcpp::Client<custom_interfaces::srv::Acc>::SharedPtr client_;
    void add(int64_t a, int64_t b, int64_t c)
    {
        // Create a request
        auto request = std::make_shared<custom_interfaces::srv::Acc::Request>();
        request->a = a;
        request->b = b;
        request->c = c;

        // Wait for the service to be available
        while (!client_->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_INFO(this->get_logger(), "Service not available, waiting again...");
            if (!rclcpp::ok())
            {
                RCLCPP_ERROR(this->get_logger(), "Interrupted while waiting for the service. Exiting.");
                return;
            }
        }

        // Process the response after the executor completes the request.
        client_->async_send_request(
            request,
            [this, request](rclcpp::Client<custom_interfaces::srv::Acc>::SharedFuture future)
            {
                response_callback(future, request);
            });
    }

private:
    void response_callback(
        rclcpp::Client<custom_interfaces::srv::Acc>::SharedFuture future,
        custom_interfaces::srv::Acc::Request::SharedPtr request)
    {
        try
        {
            auto response = future.get();
            RCLCPP_INFO(
                this->get_logger(),
                "Result of add(%ld, %ld, %ld): %ld",
                request->a,
                request->b,
                request->c,
                response->sum);
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Service call failed: %s", e.what());
        }
    }
};

int main(int argc, char *argv[])
{
    // Initialize the ROS 2 client library
    rclcpp::init(argc, argv);

    // Create a custom node
    auto node = std::make_shared<AccClientNode>();
    // Call the service with some example values
    node->add(1, 2, 3);
    node->add(4, 5, 6);
    node->add(7, 8, 9);
    // Keep the node alive until it is shut down
    rclcpp::spin(node);
    // Destroy the node
    node.reset();

    // Shutdown the ROS 2 client library
    rclcpp::shutdown();
    return 0;
}
